controllers/storageclassclaimsecret/storageclassclaimsecret_controller.go

```diff
@@ -0,0 +1,691 @@
+// +kubebuilder:rbac:groups=ceph.rook.io,resources=cephblockpools,verbs=get;list;watch;create;update;delete
+// +kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusters,verbs=get;watch;list
+// +kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=snapshot.storage.k8s.io,resources=volumesnapshotclasses,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+
+func (r *StorageClassClaimSecretReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.log = ctrllog.FromContext(ctx, "StorageClassClaim Secret", request)
+	r.ctx = ctrllog.IntoContext(ctx, r.log)
+	r.log.Info("Reconciling StorageClassClaim Secret.")
+
+	// Fetch the StorageClassClaim instance
+	r.storageClassClaimSecret = &corev1.Secret{}
+	r.storageClassClaimSecret.Name = request.Name
+	r.storageClassClaimSecret.Namespace = request.Namespace
+
+	if err := r.get(r.storageClassClaimSecret); err != nil {
+		if errors.IsNotFound(err) {
+			r.log.Info("StorageClassClaim Secret not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		r.log.Error(err, "Failed to get StorageClassClaim Secret.")
+		return reconcile.Result{}, err
+	}
+
+	r.storageClassClaim = &v1alpha1.StorageClassClaim{}
+	if unmarshalErr := json.Unmarshal(r.storageClassClaimSecret.Data["StorageClassClaim"], r.storageClassClaim); unmarshalErr != nil {
+		r.log.Error(unmarshalErr, "Failed to unmarshal StorageClassClaim Secret.")
+		return reconcile.Result{}, unmarshalErr
+	}
+
+	if r.storageClassClaim.Status.Phase == "" {
+		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimInitializing
+	}
+
+	storageClusterList := &v1.StorageClusterList{}
+	if err := r.list(storageClusterList, client.InNamespace(r.OperatorNamespace)); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch l := len(storageClusterList.Items); {
+	case l == 0:
+		return reconcile.Result{}, fmt.Errorf("no StorageCluster found")
+	case l != 1:
+		return reconcile.Result{}, fmt.Errorf("multiple StorageCluster found")
+	}
+	r.storageCluster = &storageClusterList.Items[0]
+
+	var result reconcile.Result
+	var reconcileError error
+	if storagecluster.IsOCSConsumerMode(r.storageCluster) {
+		return reconcile.Result{}, fmt.Errorf("StorageClassClaim Secret is not supported in OCSConsumer mode")
+	}
+
+	result, reconcileError = r.reconcilePhases()
+
+	if claimBytes, marshalErr := json.Marshal(r.storageClassClaim); marshalErr == nil {
+		r.storageClassClaimSecret.Data["StorageClassClaim"] = claimBytes
+	} else {
+		r.log.Error(marshalErr, "Failed to marshal StorageClassClaim Secret.")
+		return reconcile.Result{}, marshalErr
```

Comment:
I think this should be a Panic (or os.Exist), not a returned Error. 
If we cannot marshal it means we have a fundamental code error.